From 60596b1e77aeaae5f952f60549835894f14d3478 Mon Sep 17 00:00:00 2001
From: itsmanjeet <itsmanjeet1998@gmail.com>
Date: Wed, 10 Jan 2024 17:54:46 +0530
Subject: [PATCH 1/2] Rearranged menu items

- Added Launch new instance
- Use submenu to manage wnck items
---
 src/Wnck.cpp | 29 +++++++++++++++++++++++------
 1 file changed, 23 insertions(+), 6 deletions(-)

diff --git a/src/Wnck.cpp b/src/Wnck.cpp
index e4fff14..1bc054b 100644
--- a/src/Wnck.cpp
+++ b/src/Wnck.cpp
@@ -184,9 +184,25 @@ G_GNUC_END_IGNORE_DEPRECATIONS
 
 	GtkWidget* buildActionMenu(GroupWindow* groupWindow, Group* group)
 	{
-		GtkWidget* menu = (groupWindow != nullptr && !groupWindow->getState(WNCK_WINDOW_STATE_SKIP_TASKLIST)) ? wnck_action_menu_new(groupWindow->mWnckWindow) : gtk_menu_new();
+		GtkWidget* menu = gtk_menu_new ();
 		std::shared_ptr<AppInfo> appInfo = (groupWindow != nullptr) ? groupWindow->mGroup->mAppInfo : group->mAppInfo;
 
+		if (groupWindow != nullptr && !groupWindow->getState(WNCK_WINDOW_STATE_SKIP_TASKLIST)) {
+			GtkWidget* wnck_action_items = gtk_menu_item_new_with_label(_("Manage Window"));
+			gtk_menu_item_set_submenu(GTK_MENU_ITEM(wnck_action_items), wnck_action_menu_new(groupWindow->mWnckWindow));
+			gtk_menu_shell_append(GTK_MENU_SHELL(menu), wnck_action_items);
+		}
+
+		gtk_menu_shell_append(GTK_MENU_SHELL(menu), gtk_separator_menu_item_new());
+
+		GtkWidget* launchNewInstance = gtk_menu_item_new_with_label(_("Launch New Instance"));
+		gtk_menu_shell_append(GTK_MENU_SHELL (menu), launchNewInstance);
+
+		g_signal_connect (G_OBJECT (launchNewInstance), "activate",
+			G_CALLBACK (+[](GtkMenuItem* menu_item, AppInfo* _appInfo) -> void {
+					_appInfo->launch();
+				}), appInfo.get());
+
 		if (!appInfo->path.empty())
 		{
 			const gchar* const* actions = appInfo->get_actions();
@@ -221,6 +237,7 @@ G_GNUC_END_IGNORE_DEPRECATIONS
 				gtk_check_menu_item_set_active(GTK_CHECK_MENU_ITEM(pinToggle), group->mPinned);
 				gtk_menu_shell_prepend(GTK_MENU_SHELL(menu), gtk_separator_menu_item_new());
 
+
 				gchar* program = g_find_program_in_path("exo-desktop-item-edit");
 				if (program != nullptr)
 				{
@@ -239,11 +256,11 @@ G_GNUC_END_IGNORE_DEPRECATIONS
 					}),
 					group);
 
-				g_signal_connect(G_OBJECT(editLauncher), "activate",
-					G_CALLBACK(+[](GtkMenuItem* menuitem, AppInfo* _appInfo) {
-						_appInfo->edit();
-					}),
-					appInfo.get());
+//				g_signal_connect(G_OBJECT(editLauncher), "activate",
+//					G_CALLBACK(+[](GtkMenuItem* menuitem, AppInfo* _appInfo) {
+//						_appInfo->edit();
+//					}),
+//					appInfo.get());
 
 				if (group->mWindowsCount > 1)
 				{
-- 
2.42.0

